File: neetcode all and misc questions/Arrays and Strings/unique_length_3_palindromic_subsequences.py
class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        """
        
        c a b c a
        start from the mid? at b
        
           a b c
        neither side is a palindrome, move one more to the right
        a b a, one palindrome found, put it in set so we don't double count it
        
        c a b c a
        
        can also try moving one more left
        c b c
        
        c a b d c
        """
        """
        Brute Force
        """
        # unique = set()
        # for i in range(1, len(s)-1):
        #     l = i - 1
        #     r = i + 1
        #     while l >= 0 and r < len(s):
        #         leftVal = s[l]
        #         rightVal = s[r]
        #         newL = l
        #         newR = r
        #         if leftVal == rightVal:
        #             unique.add(s[l] + s[i] + s[r])
        #         # continuously check the right side
        #         while (newR < len(s)-1):
        #             newR += 1
        #             if s[newR] == s[l]:
        #                 unique.add(s[l] + s[i] + s[newR])
        #         # continuously check the right side
        #         while (newL > 0):
        #             newL -= 1
        #             if s[newL] == s[r]:
        #                 unique.add(s[newL] + s[i] + s[r])       
        #         l -= 1
        #         r += 1
        # return len(unique)
        """
        Optimal Solution (Neetcode)
        Key Concepts:
        1) When doing a string problems that only use letters of the alphabet,
        always remember that there's 26 characters max in the alphabet and see if you can use
        that fact to solve the problem.
        2) In palindromes of length 3, the character to the left of the middle and the character to the right of the middle have to be the same. We can use this fact by simply looping through and treating each
        s[i] as a "middle", and then seeing if it's possible that any character on the left of middle,
        and any character to the right of middle can be the same.
        
        Time: O(26 * N)
        
        example:
        
        s = "bbcbaba"
        
        answer = 4
            
        right = {a: 2, b: 4, c: 1}
        left = set()
        res = set()
        
        i = 0
        b 
        decrement b from right
        {a: 2, b: 3, c: 1}
        
        from a ... z, 
        because left is empty, we can't make a length three palindrome here,
        so add b to "left" and go onto the next iteration
        
        i = 1
        b
        left = {b}
        res = set()
        decrement b from right
        right = {a: 2, b: 2, c: 1}
        
        from a ... z,
        b is present in both left and right, so we can make a length three palindrome here,
        add bbb to our res set
        res = {bbb}
        
        adds b to left, it already exists, move onto the next iteration
        
        i = 2
        c
        decrement c from right, and pops c out since the value becomes 0 
        right = {a: 2, b: 2}
        
        from a ... z,
        b is present in both left and right, so we can make a length three palindrome,
        add bcb to our res set
        res = {bbb, bcb}
        
        ...
        
        
        
        """
        from collections import Counter
        res = set()
        # initially left side is empty
        left = set()
        # right side contains all the characters of S
        right = Counter(s)
        for i in range(len(s)):
            # middle of our palindrome
            mid = s[i]
            # because we're evaluating this character in the middle,
            # we want to remove this from the possible list of chars on the right 
            right[mid] -= 1
            if right[mid] == 0:
                right.pop(mid)
            
            for j in range(26):
                # ASCII math, this will convert our index
                # to it's appropriate alphabet character
                # based on the fact that there's 26 possible letters,
                # so 0 = a, 1 = b, 2 = c, 25 = z, etc
                c = chr(ord("a") + j ) 
                # if c in left and c in right, that means a 3 length palindrome can be created
                # since the characters would be the same here
                if c in left and c in right:
                    res.add(c + mid + c)
            
            # after we evaluate the middle, it is now going to be in the left portion
            left.add(mid)
        return len(res)

File: node_js_ds_and_algos_implementations/arrays and strings/hashmaps/unique_length_3_palindromic_subsequences.js
/**
 * @param {string} s
 * @return {number}
 */
var countPalindromicSubsequence = function(s) {
    /*
    Most efficient solution
    O(26 * N) time
    O(N) space
    Similar to the better solution below,
    but rather than repeatedly iterating through left and right, we can track characters that we've already seen
    using a hashmap
    Use a hashmap that represents the counts of all characters on the right of the current index
    Use a set that represents all characters that were formerly on the right, but are now on the "left", indicating
    that we've seen it
    Use another set to check for duplicate palindromes
    Neetcode: https://youtu.be/3THUt0vAFLU
    
    Example:
    s = "bbcbaba", answer should be 4
    
    set = {}
    left = {}
    right = {b: 4, a:2, c:1}
    
    1st iteration i = 0
    -----------------------
    s[0] is b
    right[b]-- (right[b] is now 3)
    
    because we're still at i = 0, we can't find a character to the left that matches a character on the right,
    since left is still empty
    
    add b to left 
    
    set = {}
    left = {b}
    right = {b: 3, a:2, c:1}
    
    2nd iteration i = 1
    -----------------------
    s[1] = b
    right[b]-- (right[b] is now 2)
    
    searching on both the left and right for the same letter, you can see that b exists
    on both the left and right, forming "bbb". Add "bbb" to the set. No other characters exists on both sides
    
    add b to left (b already exists in the set so it's not added)
    
    set = {bbb}
    left = {b}
    right = {b: 2, a:2, c:1}
    
    3rd iteration i = 2
    -----------------------
    s[2] = c
    right[c]-- (right[c] is now 0, so remove c as a key)
    
    searching on both the left and right for the same letter, you can see that b exists on both sides
    forming the palindrome "bcb", add to set. No other character exists on both sides
    
    add c to left
    
    set = {bbb, bcb}
    left = {b, c}
    right = {b: 2, a:2}
    
    4th iteration i = 3
    -----------------------
    s[3] = b
    right[b]-- (right[b] is now 1)
    
    we can see that b exists on the left and right, but we've already added bbb to our set so ignore it
    no other character exists on both sides
    
    add b to left (b is already added so ignore)
    
    set = {bbb, bcb}
    left = {b, c}
    right = {b: 1, a: 2}
    
    5th iteration i = 4
    ----------------------------
    s[4] = a
    right[a]-- (right[a] is now 1)
    
    we can see that b exists on the left and right, forming the palindrome bab, add to our set
    
    add a to left 
    
    set = {bbb, bcb, bab}
    left = {b, c, a}
    right = {b: 1, a: 1}
    
    6th iteration i = 5
    --------------------------------
    s[5] = b
    right[b]-- (right[b] is now 0, remove this key)
    
    we can see that b exists on the left and right, a exists on the left and right.
    we've added bab but not aba, so add aba
    
    add b to left (already been added, so ignore)
    
    set = {bbb, bcb, bab, aba}
    left = {b, c, a}
    right = {a: 1}
    
    7th iteration i = 6
    --------------------------
    s[6] = a
    right[a]-- (right[a] is now 0, remove this key)
    there's nothing on the right side that exists on the left since the right side is now empty
    
    adds a to the left (already been added, so ignore)
    
    ------------------------------------
    At the end, there's four unique palindromes in the set, so return 4
    
    */
    let set = new Set()
    let left = new Set()
    let right = {}
    for (let char of s){
        if (char in right){
            ++right[char]
        }
        else {
            right[char] = 1
        }
    }
    let letters = "abcdefghijklmnopqrstuvwxyz"
    for (let i = 0; i < s.length; ++i){
        --right[s[i]]
        if (right[s[i]] === 0){
            delete right[s[i]]
        }
        for (let letter of letters){
            if (left.has(letter) && letter in right){
                set.add(letter + s[i] + letter)
            }
        }
        left.add(s[i])
    }
    return set.size
    /*
    Better solution:
    O(26 * N^2) solution
    Key observation:
    For length 3 palindromes, the middle character can be any character from a to z, while the two surrounding 
    characters always have to be the same.
    
        We still keep a set to check for duplicate palindromes
        z a a b c a b z
        
        at each index, this would be the "middle" of the palindrome
        then we can slice everything from 0 to i (left side) and i + 1 to s.length (right side),
        then loop through all letters a to z to check if the letter exists in both the left and right side.
        Also check if the palindrome (letter + s[i] + letter) already exists in our set
    */
    /*
    if (s.length < 3){
        return 0
    }
    let counter = 0
    let set = new Set()
    let letters = "abcdefghijklmnopqrstuvwxyz"
    for (let i = 1; i < s.length-1; ++i){
        let left = s.slice(0, i)
        let right = s.slice(i+1, s.length)
        for (let letter of letters){
            let palindrome = letter + s[i] + letter
            if (left.includes(letter) && right.includes(letter) && !set.has(palindrome)){
                set.add(palindrome)
                ++counter
            }
        }
    }
    return counter
    */
    
    /*
    O(N^3) brute force solution
    1) generate every unique 3 length subsequence in the string
    2) check if each subsequence is a palindrome
    */
    /*
    var isPalindrome = function(s){
        let l = 0
        let r = s.length-1
        while (l <= r){
            if (s[l] !== s[r]){
                return false
            }
            l++
            r--
        }
        return true
    }
    if (s.length >= 3){
        let set = new Set()
        for (let i = 0; i < s.length; ++i){
            for (let j = i+1; j < s.length; ++j){
                for (let k = j+1; k < s.length; ++k){
                    let subsequence = s[i] + s[j] + s[k]
                    set.add(subsequence)
                }
            }
        }
        let counter = 0
        for (let word of set){
            if (isPalindrome(word)){
                ++counter
            }
        }
        return counter
    }
    return 0
    */
};